libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/TextComposer.kt

```diff
@@ -105,319 +105,404 @@ fun TextComposer(
-                        composerMode = composerMode,
-                        modifier = it
-                    )
-                },
-            )
         }
     }
-
-    SoftKeyboardEffect(composerMode, onRequestFocus) {
-        it is MessageComposerMode.Special
-    }
-
-    SoftKeyboardEffect(showTextFormatting, onRequestFocus) { it }
 }
 
 @Composable
 private fun TextInput(
     state: RichTextEditorState,
+    subcomposing: Boolean,
     placeholder: String,
-    roundedCorners: RoundedCornerShape,
-    bgColor: Color,
+    composerMode: MessageComposerMode,
+    onResetComposerMode: () -> Unit,
     modifier: Modifier = Modifier,
     onError: (Throwable) -> Unit = {},
 ) {
-    val minHeight = 42.dp.applyScaleUp()
-    val defaultTypography = ElementTheme.typography.fontBodyLgRegular
-    Box(
+    val bgColor = ElementTheme.colors.bgSubtleSecondary
+    val borderColor = ElementTheme.colors.borderDisabled
+    val roundedCorners = textInputRoundedCornerShape(composerMode = composerMode)
+
+    Column(
         modifier = modifier
-            .heightIn(min = minHeight)
-            .background(color = bgColor, shape = roundedCorners)
-            .padding(
-                PaddingValues(
+            .clip(roundedCorners)
+            .border(0.5.dp, borderColor, roundedCorners)
+            .background(color = bgColor)
+            .requiredHeightIn(min = 42.dp.applyScaleUp())
+            .fillMaxSize(),
+    ) {
+        if (composerMode is MessageComposerMode.Special) {
+            ComposerModeView(composerMode = composerMode, onResetComposerMode = onResetComposerMode)
+        }
+        val defaultTypography = ElementTheme.typography.fontBodyLgRegular
+        Box(
+            modifier = Modifier
+                .padding(
                     top = 4.dp.applyScaleUp(),
                     bottom = 4.dp.applyScaleUp(),
                     start = 12.dp.applyScaleUp(),
                     end = 42.dp.applyScaleUp()
                 )
-            )
-            .testTag(TestTags.richTextEditor),
-        contentAlignment = Alignment.CenterStart,
-    ) {
+                .testTag(TestTags.richTextEditor),
+            contentAlignment = Alignment.CenterStart,
+        ) {
+            // Placeholder
+            if (state.messageHtml.isEmpty()) {
+                Text(
+                    placeholder,
+                    style = defaultTypography.copy(
+                        color = ElementTheme.colors.textSecondary,
+                    ),
+                    maxLines = 1,
+                    overflow = TextOverflow.Ellipsis,
+                )
+            }
 
-        // Placeholder
-        if (state.messageHtml.isEmpty()) {
-            Text(
-                placeholder,
-                style = defaultTypography.copy(
-                    color = ElementTheme.colors.textSecondary,
+            RichTextEditor(
+                state = state,
+                registerStateUpdates = !subcomposing,
```

Comment:
Disable most of the editor functionality if it's just being measured for a subcomposition. This prevents it gaining focus and mutating the state.

Reply:
OK, thanks. Maybe add this comment also in the code?